### PR TITLE
feat: support recursive download in dfget

### PR DESCRIPTION
**Pull request by jim3ma**
Signed-off-by: Jim Ma <majinjing3@gmail.com>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist:

- [ ] Code compiles correctly.
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Review by gaius-qi — DISMISSED** [review]
LGTM

File: client/config/dfget.go [hunk 1]
```diff
@@ -181,13 +184,22 @@ func (cfg *ClientOption) checkOutput() error {
...
-	if err := MkdirAll(dir, 0777, basic.UserID, basic.UserGroup); err != nil {
+	outputDir, _ := path.Split(cfg.Output)
+	if err := MkdirAll(outputDir, 0777, basic.UserID, basic.UserGroup); err != nil {
 		return err
 	}
 
-	if f, err := os.Stat(cfg.Output); err == nil && f.IsDir() {
-		return fmt.Errorf("path[%s] is directory but requires file path", cfg.Output)
+	f, err := os.Stat(cfg.Output)
+	// when recursive download, need a directory
```
Comment (244372610) [inline client/config/dfget.go:193]: if err != nil {
  return nil
}
这样就不用在下面的代码中判断 err == nil了，代码结构是否会更清晰些

File: client/config/dfget.go [hunk 2]
```diff
@@ -181,13 +184,22 @@ func (cfg *ClientOption) checkOutput() error {
 		cfg.Output = absPath
 	}
 
-	dir, _ := path.Split(cfg.Output)
-	if err := MkdirAll(dir, 0777, basic.UserID, basic.UserGroup); err != nil {
+	outputDir, _ := path.Split(cfg.Output)
+	if err := MkdirAll(outputDir, 0777, basic.UserID, basic.UserGroup); err != nil {
```
Comment (244372610) [inline client/config/dfget.go:188]: 不建议调用MkdirAll,我的想法是需要先判断cfg.Output的父目录是否存在，如果不存在就报错（类似wget），如果存在则创建outputDir以及目录下的其他文件和文件夹。下载普通文件的时候也类似，比如先确定目标文件路径是否存在，如果不存在报错，存在的话正常下载

File: client/dfget/dfget.go [hunk 3]
```diff
@@ -59,7 +61,11 @@ func Download(cfg *config.DfgetConfig, client daemonclient.DaemonClient) error {
 
 	go func() {
 		defer cancel()
-		downError = download(ctx, client, cfg, wLog)
+		if cfg.Recursive {
+			downError = recursiveDownload(ctx, client, cfg)
```
Comment (244372610) [inline client/dfget/dfget.go:65]: 建议这里的if/else判断放到download方法中，在download方法中获取到urls列表，对应到之前的普通文件下载url切片长度为1，这样客户端的代码量会少一些。这条只是建议

File: client/config/dfget.go (hunk 1 above)
Comment (jim3ma) [reply client/config/dfget.go:193]: 还要检查权限，不能直接返回

**Review by gaius-qi — APPROVED** [review]
LGTM

**Review by 244372610 — APPROVED** [review]

**Pull request merged**